lexicon/conversion.py

```diff
@@ -280,3 +279,109 @@ def run(self):
         with uopen(self.out_bliss_lexicon.get_path(), "w") as lexicon_file:
             lexicon_file.write('<?xml version="1.0" encoding="utf-8"?>\n')
             lexicon_file.write(ET.tostring(lex.to_xml(), "unicode"))
+
+
+class SpellingConversionJob(Job):
+    """Convert lexicon to a new one with other regional spellings
+    e.g. US -> GB spelling
+    """
+
+    def __init__(
+        self,
+        bliss_lexicon,
+        orth_mapping_file,
+        reverse_mapping=False,
+        mapping_delimiter=" ",
+    ):
+        """
+        :param Path bliss_lexicon:
+            input lexicon
+        :param str orth_mapping_file:
+            orthography mapping file, .json .json.gz .txt .gz
+            in case of text file, one can adjust mapping_delimiter
+        :param bool reverse_mapping:
+            reverse/flip the mapping
+        :param str mapping_delimiter:
+            delimiter of source and target orths
+            if mapping is provided with a plain text file
+        """
+
+        self.set_vis_name("Convert Between Regional Orth Spellings")
+
+        self.bliss_lexicon = bliss_lexicon
+        self.orth_mapping_file = orth_mapping_file
+        self.reverse_mapping = reverse_mapping
+        self.mapping_delimiter = mapping_delimiter
+
+        self.out_bliss_lexicon = self.output_path("lexicon.xml.gz")
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def run(self):
+
+        # load mapping from json or plain text file
+        is_json = self.orth_mapping_file.endswith(".json")
+        is_json = is_json or self.orth_mapping_file.endswith(".json.gz")
+        if is_json:
+            with uopen(self.orth_mapping_file, "rt") as f:
+                mapping = json.load(f)
+            if self.reverse_mapping:
+                mapping = {v: k for k, v in mapping.items()}
+        else:
+            mapping = dict()
+            with uopen(self.orth_mapping_file, "rt") as f:
+                for line in f:
+                    line = line.strip()
+                    if not line or line.startswith("#"):
+                        continue
+                    source, target = line.split(self.mapping_delimiter)
+                    if self.reverse_mapping:
+                        source, target = target, source
+                    mapping[source] = target
+
+        # load input lexicon and build orthography to lemma dict
+        lex = lexicon.Lexicon()
+        lex.load(self.bliss_lexicon.get())
+        orth2lemma = {o: lemma for lemma in lex.lemmata for o in lemma.orth}
```

Comment:
if two lemmata have the same orth, second one will overwrite the first and first one will not be mapped. Not sure if this occurs in practice

Reply:
the key of the mapping is orth token, as long as orths in the original lemma are all unique, the mapping is there.

Reply:
hmm, you are right! in what case will this happen? if this is affecting the idea of converting spelling, then I will better make the value a list